### PR TITLE
@tus/server: refactor and improve request validator

### LINKED ISSUE #378: _invalidTusResumableHeader, _invalidUploadLengthHeader, _invalidUploadMetadataHeader using tus-js-client
I am using the tus-js-client version ^3.0.1 and the @tus/server ^1.0.0-beta.2 with @tus/gcs-store  ^1.0.0-beta.1. I didn't change any config. Inside my react component I am using: 

```javascript
    const [file, setFile] = useState();

    const handleDropZoneDrop = useCallback(
        async (_dropFiles, acceptedFiles, _rejectedFiles) => {
            setFile(acceptedFiles ? acceptedFiles[0] : null);
            onVideoChanged(acceptedFiles[0]);
        },
        [],
    );

    const { upload, setUpload, isSuccess, error, remove } = useTus();

    useEffect(() => {
            if (!file) {
                return;
            }

            setUpload(file, {
                endpoint: '/api/upload/addvideo',
                metadata: {
                    filename: `${id}.mp4`,
                    filetype: 'video/mp4',
                },
            });
        },
        [file, setUpload]
    );

    useEffect(() => {
        if (!upload) {
            return;
        }

        // Start to upload the file.
        upload.start();
    }, [upload]);

```

My node.js server is using: 

```javascript
const server = new Server({
  path: 'add-video',
  datastore: new GCSStore({
    keyFilename: join(__dirname, "./backend/google-cloud/<filename>.json"),
    projectId: "<projectId>",
    bucket: '<bucketName>',
  }),
})

uploadApp.all('*', server.handle)
app.use('/api/upload/*', (req, res) => {
  return uploadApp(req, res);
})

```
My headers are:

```
{
   connection: 'close',
   host: '4648-194-96-91-120.eu.ngrok.io',
   'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36',
   'content-length': '0',
  accept: '*/*',
   'accept-encoding': 'gzip, deflate, br',
   'accept-language': 'en-GB,en-US;q=0.9,en;q=0.8',
   cookie: 'abuse_interstitial=4648-194-96-91-120.eu.ngrok.io',
   origin: 'https://4648-194-96-91-120.eu.ngrok.io',
  referer: 'https://4648-194-96-91-120.eu.ngrok.io/?embedded=1&hmac=6d1f98d74b11b8b276a90f146a1f6df3ae6767ba680c06f3da58d361863abef6&host=c2hvcGFibGUtc3RyZWFtLm15c2hvcGlmeS5jb20vYWRtaW4&locale=de&se
                                   ssion=8f26f80c28a3ecdcdae1377e2346161003fd0eca8ad37d56ad9894197b6c37a4&shop=shopable-stream.myshopify.com&timestamp=1673116133',
   'sec-ch-ua': '"Not?A_Brand";v="8", "Chromium";v="108", "Google Chrome";v="108"',
   'sec-ch-ua-mobile': '?0',
   'sec-ch-ua-platform': '"macOS"',
   'sec-fetch-dest': 'empty',
   'sec-fetch-mode': 'cors',
   'sec-fetch-site': 'same-origin',
   'tus-resumable': '1.0.0',
   'upload-length': '65233579',
   'upload-metadata': 'filename NjNiNzMwMmYwNzliZDIyZmFlNjAyOTNhLm1wNA==,filetype dmlkZW8vbXA0',
   'x-forwarded-for': '194.96.91.120',
   'x-forwarded-proto': 'https'
 }
```

The exception I am getting is the following one: 

```
backend    | [Function: _invalidTusResumableHeader]
backend    | [Function: _invalidUploadLengthHeader]
backend    | [Function: _invalidUploadMetadataHeader]
backend    | /Users/lunaticcoding/Documents/einzelunternehmen/shopable_stream/shopable-stream-shopify/web/node_modules/@tus/server/dist/validators/RequestValidator.js:76
backend    |         return this[method](header_value);
backend    |                            ^
```

I am new to the tus protocol. Can someone help me please?

### PR DESCRIPTION
**Pull request by Murderlon**
Closes #378

- Refactor to be less fragile and fix a crash when we want to parse a unknown header. 
- Fix number validation
- Fix metadata parsing to also fail on `'/t/n'` and `''`.
  - Because of this a test had to be removed. If `Upload-Metadata` is invalid, it will already be caught early in the `handle` function. Before we send a more meaningful message in the post handler. I think it's better to catch it early and if we want better messages, such as the spec message per failed header, then that's a separate issue to tackle.

### CODE REVIEW

**Comment by mitjap** [comment]
I think it is confusing to have a list of validators that in fact check for invalidity.

**Comment by Murderlon** [comment]
Good point. It's only in the internal variable name though. I kept the workings the same as before, which was invalidity. Might be better yes to flip it around.

File: packages/server/src/validators/HeaderValidator.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    function (value) {
+      return Number.parseInt(value as string, 10) >= 0
+    },
+  ],
+  [
+    // The Upload-Length request and response header indicates the size of the entire upload in bytes.
+    // The value MUST be a non-negative integer.
+    'upload-length',
+    function (value) {
+      return Number.parseInt(value as string, 10) > 0
```
Comment (mitjap) [inline packages/server/src/validators/HeaderValidator.ts:20]: Also allow 0.

```suggestion
      return Number.parseInt(value as string, 10) >= 0
```

File: packages/server/src/validators/HeaderValidator.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    },
+  ],
+  [
+    // The Upload-Defer-Length request and response header indicates that the size of the upload
+    // is not known currently and will be transferred later.
+    // Its value MUST be 1. If the length of an upload is not deferred, this header MUST be omitted.
+    'upload-defer-length',
+    function (value) {
+      return Number.parseInt(value as string, 10) === 1
```
Comment (mitjap) [inline packages/server/src/validators/HeaderValidator.ts:29]: What about a simple string comparison?
```suggestion
      return value === '1'
```

File: packages/server/src/validators/HeaderValidator.ts (hunk 1 above)
Comment (mitjap) [reply packages/server/src/validators/HeaderValidator.ts:20]: I don't think `Number.parseInt` is appropriate for any validation. It parses many non-numerical inputs as valid numbers. Maybe simple regex?
```
const regex = /^[1-9][0-9]*$/
return value === '0' || regex.test(value);
```

Example of invalid input that would be parsed as valid number:
`"0100" -> 100`
`"0asd100" -> 100`
`"1asd100" -> 1`

File: packages/server/test/HeaderValidator.test.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  describe('upload-offset', () => {
+    it('should validate a number', (done) => {
+      const value = '1234'
+      assert.equal(validateHeader('upload-offset', value), true)
+      done()
+    })
+
+    it('should invalidate a negative number', (done) => {
+      const value = '-4'
+      assert.equal(validateHeader('upload-offset', value), false)
```
Comment (mitjap) [inline packages/server/test/HeaderValidator.test.ts:16]: Maybe add some more complex values that are falsely validated as number?

**Comment by mitjap** [comment]
Sorry for nagging on such minor refactor :)

**Comment by Murderlon** [comment]
> Sorry for nagging on such minor refactor :)

Not all all! I went too quick about this refactor, not thinking enough about potential other improvements.

File: packages/server/src/validators/HeaderValidator.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      return n !== Infinity && String(n) === value && n >= 0
+    },
+  ],
+  [
+    // The Upload-Length request and response header indicates the size of the entire upload in bytes.
+    // The value MUST be a non-negative integer.
+    'upload-length',
+    function (value) {
+      const n = Number(value)
+      return n !== Infinity && String(n) === value && n >= 0
```
Comment (mitjap) [inline packages/server/src/validators/HeaderValidator.ts:22]: I think this version also parses non-integer values.

File: packages/server/test/HeaderValidator.test.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  })
+
+  describe('upload-length', () => {
+    it('should validate a number', (done) => {
+      const value = '1234'
+      assert.equal(validateHeader('upload-length', value), true)
+      done()
+    })
+
+    it('should invalidate a number < 1', (done) => {
```
Comment (mitjap) [inline packages/server/test/HeaderValidator.test.ts:36]:
```suggestion
    it('should invalidate a number < 0', (done) => {
```

File: packages/server/test/HeaderValidator.test.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    it('should validate 1', (done) => {
+      const value = '1'
+      assert.equal(validateHeader('upload-defer-length', value), true)
+      done()
+    })
+
+    it('should invalidate a number !== 1', (done) => {
+      assert.equal(validateHeader('upload-defer-length', '0'), false)
+      assert.equal(validateHeader('upload-defer-length', '1234'), false)
+      assert.equal(validateHeader('upload-defer-length', '-1'), false)
```
Comment (mitjap) [inline packages/server/test/HeaderValidator.test.ts:60]: Added some more tests that might fail if `Number.parseInt(value) === 1` is used.

```suggestion
      assert.equal(validateHeader('upload-defer-length', '-1'), false)
      assert.equal(validateHeader('upload-defer-length', '+1'), false)
      assert.equal(validateHeader('upload-defer-length', ' 1 '), false) // test leading and trailing whitespaces
```

File: packages/server/test/HeaderValidator.test.ts [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    it('should invalidate a number < 1', (done) => {
+      assert.equal(validateHeader('upload-length', '-1'), false)
+      done()
+    })
+
+    it('should invalidate a non number', (done) => {
+      assert.equal(validateHeader('upload-length', 'hello'), false)
+      assert.equal(validateHeader('upload-length', '0100'), false)
+      assert.equal(validateHeader('upload-length', '0asd100'), false)
+      assert.equal(validateHeader('upload-length', '1asd100'), false)
```
Comment (mitjap) [inline packages/server/test/HeaderValidator.test.ts:45]:
```suggestion
      assert.equal(validateHeader('upload-length', '1asd100'), false)
      assert.equal(validateHeader('upload-length', '1.3'), false)
      assert.equal(validateHeader('upload-length', '-0'), false)
      assert.equal(validateHeader('upload-length', '+0'), false)
      assert.equal(validateHeader('upload-length', 'NaN'), false)
      assert.equal(validateHeader('upload-length', '+Infinity'), false)
```

**Review by mitjap — CHANGES_REQUESTED** [review]

**Review by mitjap — APPROVED** [review]

**Pull request merged**